content/en/blog/_posts/2021-07-21-memory-manager-beta.md

```diff
@@ -0,0 +1,106 @@
+- The amount of reserved memory for the resource type must be equal to [NodeAllocatable](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/) (`kube-reserved + system-reserved + eviction-hard`) for the resource. You can read more about 
+
+![ReservedMemory
+](/images/blog/2021-07-21-memory-manager-beta/ReservedMemory.png "ReservedMemory")
+
+## Current limitations
+
+The 1.22 release and promotion to the beta brings along some enhancements and fixes, but the **MemoryManager** still has several limitations.
+
+### Single vs Cross NUMA node allocation
+
+The NUMA node can not have both single and cross NUMA node allocations. When the container memory is pinned to two or more NUMA nodes, we can not know from which NUMA node the container will consume the memory.
+
+![SingleCrossNUMAAllocation
+](/images/blog/2021-07-21-memory-manager-beta/SingleCrossNUMAAllocation.png "SingleCrossNUMAAllocation")
+
+1. The `container1` started on the NUMA node 0 and requests *5Gi* of the memory but currently is consuming only *3Gi* of the memory.
+2. The `container2` the memory request is *10Gi*, and none single NUMA node can satisfy it.
+3. The `container2` consumes *3.5Gi* of the memory from the NUMA node 0, but once the `container1` will require more memory, it will not have it, and the kernel will kill one of the containers with the **OOM** error.
+
+To prevent such issues, the **MemoryManager** will fail the admission of the `container2` until the machine, will have two NUMA nodes without a single NUMA node allocation.
+
+### Works only for guaranteed pods
+
+The **MemoryManager** can not guarantee memory allocation for burstable pods, also when the burstable pod has specified equal memory limit and request.
+Let's assume you have two burstable pods, `pod1` has containers with equal memory request and limits, and `pod2` has containers only with the memory request. We want to guarantee memory allocation for the `pod1`. Both pods have the same **OOM** score, so once the kernel will not have enough memory, it can kill `pod1`.
+
+### Memory fragmentation
+
+The sequence of containers that start and stop can fragment the memory on NUMA nodes. Currently, we do not have any components under the kubelet that can balance pods and defragment memory back.
+
+## Future work for the **MemoryManager**
+
+We do not want to stop with the current state of the **MemoryManager** and looking for places for improvements.
+
+### Make the MemoryManager allocation algorithm smarter
+
+The current algorithm ignores distances between NUMA nodes during the calculation of the allocation. We can get better performance when the **MemoryManager** will use the closest NUMA nodes for cross NUMA node allocation.
+
+### Reduce the number of admission errors
+
+The default Kubernetes scheduler not aware of the node's NUMA topology, and it can be a reason for many admission errors during the pod start.
```

Comment:
```suggestion
The default Kubernetes scheduler is not aware of the node's NUMA topology, and it can be a reason for many admission errors during the pod start.
```

Reply:
done